src/main/java/software/amazon/awssdk/crt/http/HttpConnection.java

```diff
@@ -0,0 +1,218 @@
+    private static final String HTTP = "http";
+    private static final String HTTPS = "https";
+
+    private final ClientBootstrap clientBootstrap;
+    private final SocketOptions socketOptions;
+    private final TlsContext tlsContext;
+    private final URI uri;
+    private final HttpConnectionEvents userConnectionCallbacks;
+
+    private ConnectionState connectionState = ConnectionState.DISCONNECTED;
+    private CompletableFuture<Boolean> connectedFuture;
+    private AsyncCallback connectAck;
+    private AsyncCallback disconnectAck;
+
+    public enum ConnectionState {
+        DISCONNECTED, CONNECTING, CONNECTED, SHUTTING_DOWN, SHUT_DOWN
+    }
+
+    /**
+     * Constructs a new HttpConnection. Connections are reusable after being disconnected.
+     * @param uri Must be non-null
+     * @throws HttpException If httpClient is null
+     */
+    public HttpConnection(URI uri) throws HttpException, CrtRuntimeException {
+        this(uri, new ClientBootstrap(1), new SocketOptions(), new TlsContext(), null);
+        own(this.clientBootstrap);
+        own(this.socketOptions);
+        own(this.tlsContext);
+    }
+
+    /**
+     * Constructs a new HttpConnection. Connections are reusable after being disconnected.
+     * @param uri Must be non-null
+     * @param callbacks Optional handler for connection interruptions/resumptions
+     * @throws HttpException If httpClient is null
+     */
+    public HttpConnection(URI uri, ClientBootstrap bootstrap, SocketOptions socketOptions, TlsContext tlsContext,
+                          HttpConnectionEvents callbacks) throws HttpException {
+
+        if (uri == null) {  throw new IllegalArgumentException("URI must not be null"); }
+        if (uri.getHost() == null) { throw new IllegalArgumentException("URI does not have a Host name"); }
+        if (bootstrap == null || bootstrap.isNull()) {  throw new IllegalArgumentException("ClientBootstrap must not be null"); }
+        if (socketOptions == null || socketOptions.isNull()) { throw new IllegalArgumentException("SocketOptions must not be null"); }
+        if ((HTTPS.equals(uri.getScheme()) || uri.getPort() == 443) && tlsContext == null) { throw new IllegalArgumentException("TlsContext must not be null if https is used"); }
+
+        this.uri = uri;
+        this.clientBootstrap = bootstrap;
+        this.socketOptions = socketOptions;
+        this.tlsContext = tlsContext;
+        this.userConnectionCallbacks = callbacks;
+    }
+
+    /**
+     * Connects to the Http Endpoint
+     * @return Future indicating when
+     */
+    public synchronized CompletableFuture<Boolean> connect() throws CrtRuntimeException {
+        if (!isNull()) {
+            return connectedFuture;
+        }
+
+        String endpoint = uri.getHost();
+        int port = uri.getPort();
+
+        /* Pick a default port if one wasn't set in the URI */
+        if (port == -1) {
+            port = HTTP.equals(uri.getScheme()) ? 80 : 443;
```

Comment:
NP: default port number constants and up on L84.

Reply:
Fixed